pkg/ipam/allocator/multipool/allocator.go

```diff
@@ -5,131 +5,66 @@ package multipool
 	"github.com/cilium/cilium/pkg/ipam/allocator"
 	cilium_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
 )
 
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "ipam-allocator-multi-pool")
+
 // Allocator implements allocator.AllocatorProvider
 type Allocator struct {
 	poolAlloc *PoolAllocator
 }
 
-type poolSpec struct {
-	ipv4CIDRs    []string
-	ipv6CIDRs    []string
-	ipv4MaskSize int
-	ipv6MaskSize int
+func (a *Allocator) Init(ctx context.Context) error {
+	a.poolAlloc = NewPoolAllocator()
+	return nil
 }
 
-const (
-	poolKeyIPv4CIDRs    = "ipv4-cidrs"
-	poolKeyIPv4MaskSize = "ipv4-mask-size"
-	poolKeyIPv6CIDRs    = "ipv6-cidrs"
-	poolKeyIPv6MaskSize = "ipv6-mask-size"
-)
-
-// parsePoolSpec parses a pool spec string in the form
-// "ipv4-cidrs:172.16.0.0/16,172.17.0.0/16;ipv4-mask-size:24".
-func parsePoolSpec(poolString string) (poolSpec, error) {
-	fields := strings.Split(strings.ReplaceAll(poolString, " ", ""), ";")
-
-	var pool poolSpec
-	for _, field := range fields {
-		key, value, ok := strings.Cut(field, ":")
-		if !ok {
-			return pool, fmt.Errorf("invalid number of key delimiters in pool spec %s", poolString)
-		}
-		switch key {
-		case poolKeyIPv4CIDRs:
-			pool.ipv4CIDRs = strings.Split(value, ",")
-			// TODO: validate individual CIDRs?
-		case poolKeyIPv4MaskSize:
-			mask, err := strconv.Atoi(value)
-			if err != nil {
-				return pool, fmt.Errorf("invalid value for key %q: %w", poolKeyIPv4MaskSize, err)
-			}
-			pool.ipv4MaskSize = mask
-		case poolKeyIPv6CIDRs:
-			pool.ipv6CIDRs = strings.Split(value, ",")
-		case poolKeyIPv6MaskSize:
-			mask, err := strconv.Atoi(value)
-			if err != nil {
-				return pool, fmt.Errorf("invalid value for key %q: %w", poolKeyIPv4MaskSize, err)
-			}
-			pool.ipv6MaskSize = mask
-		}
-	}
-	return pool, nil
+func (a *Allocator) Start(ctx context.Context, getterUpdater ipam.CiliumNodeGetterUpdater) (allocator.NodeEventHandler, error) {
+	return NewNodeHandler(a.poolAlloc, getterUpdater), nil
 }
 
-func (a *Allocator) Init(ctx context.Context) (err error) {
-	a.poolAlloc = NewPoolAllocator()
-
-	for poolName, pool := range operatorOption.Config.IPAMMultiPoolMap {
-		log.WithFields(logrus.Fields{
-			"pool-name": poolName,
-			"pool":      pool,
-		}).Debug("found pool definition")
-
-		pool, parseErr := parsePoolSpec(pool)
-		if err != nil {
-			err = errors.Join(err, fmt.Errorf("failed to parse IP pool spec for pool %s: %w", poolName, parseErr))
-			continue
+func (a *Allocator) UpsertPool(ctx context.Context, pool *cilium_v2alpha1.CiliumPodIPPool) error {
+	var ipv4CIDRs, ipv6CIDRs []string
+	var ipv4MaskSize, ipv6MaskSize int
+
+	if pool.Spec.IPv4 != nil {
+		ipv4MaskSize = int(pool.Spec.IPv4.MaskSize)
+		ipv4CIDRs = make([]string, len(pool.Spec.IPv4.CIDRs))
+		for i, cidr := range pool.Spec.IPv4.CIDRs {
+			ipv4CIDRs[i] = string(cidr)
```

Comment:
Maybe as a follow-up: Given that we pass `ipv4CIDRs` to `a.poolAlloc.UpsertPool` only and that method then converts its entries to `netip.Prefix` and `net.IPNet` later on (also ref. https://github.com/cilium/cilium/pull/25824#discussion_r1218257180.). I wonder whether we should just directly create a `[]netip.Prefix` here and pass that. This would also have the benefit that we could report CIDR format errors earlier. 

Same applies for `ipv6CIDRs`.

Reply:
Good idea. Doing it properly requires seems to require a bit more changes to `cidralloc`, so I'll do that in a follow-up PR.

Reply:
Done in https://github.com/cilium/cilium/pull/26138